### ISSUE TITLE
LDAP groups cache

### ISSUE THREAD

**Issue by abretaud** [authelia/authelia#910]
A little problem when doing the following:
 - one user logins to my webapp as user `foobar` defined in my ldap server
 - I (the admin) modify the ldap record for user `foobar`, adding him to a new ldap group `xyz`
 - user `foobar` tries to access a sub section of the webapp restricted to group `xyz`
 - the access is denied
 - the only way for `foobar` to access the page is to logout, and login again

I guess there is some sort of cache of groups used by authelia to determine if a user can access a page. Is there a way to make authelia refresh the group list for this user automatically?

**Comment by james-d-elliott** [comment]
I do not believe there is a cache, what LDAP provider are you using? I've got a feeling it's a distribution issue. Additionally can you post your authentication_backend config minus username/password, anything else sensitive?

**Comment by abretaud** [comment]
I'm using openldap, with this authelia conf (using version 4.12.0):
```
  ldap:
    url: ldap://ldap-hostname
    skip_verify: false
    base_dn: dc=example,dc=org
    username_attribute: uid
    additional_users_dn: ou=users
    users_filter: (&({username_attribute}={input})(isActive=TRUE))
    additional_groups_dn: ou=groups
    groups_filter: (&(member={dn})(objectclass=myGroupClass))
    group_name_attribute: cn
    mail_attribute: mail
    user: cn=admin,dc=example,dc=org
    password: xxxxxx
```

**Comment by nightah** [comment]
This is actually working as intended, the groups a user belongs to is determined and stored on initial login.

It is not updated again, unless the session is invalidated/a new session is created (user logs out and in again).

We have discussed implementing a cache so that if groups are updated in LDAP that a session can be updated with the new respective groups, however, this isn't something that has been designed or implemented to date.

**Comment by james-d-elliott** [comment]
Are we happy with the following design.  During GET /api/verify (the endpoint responsible for checking auth) we check:

1. Is the provider the LDAP provider
2. Is the authorization forbidden
3. Is the domain and path modified in some way by a group

If all of these factors are true, we then update the groups in the session directly from LDAP and if the groups changed we evaluate if the user has authorization again before the final checks.

**Comment by abretaud** [comment]
Looks good to me :)

**Comment by clems4ever** [comment]
Thanks for the design proposal @james-d-elliott. We actually want to cover the other way around too, i.e., a user that should not have access to a resource anymore because it has been moved out of a group. This case is not covered in your proposal and is the most important measure imo.

What would you think about having some kind of TTL in the session and refresh the profile when the TTL is reached instead and whatever the authorization for that resource? I'd also advise having a way to always force the refresh (for most sensitive setups willing to have the most up to date information at the cost of load on the LDAP) and a way to disable it totally too to keep the same behavior as previously.

Also by `Is the domain and path modified in some way by a group` do you mean `Is the domain and path covered by an ACL rule matching the group`? I don't really get the term `modified`, can you elaborate?

**Comment by james-d-elliott** [comment]
That's a really good point. Regarding `Is the domain and path modified in some way by a group` - my refresh basically checks if there is any rule that would match the request, and has a `subject: "group:*"`

I can think on t he TTL stuff, but @nightah may have a plan in mind for that.

**Comment by nightah** [comment]
Yeah I think a configurable TTL makes the most sense per our dated discussions on this @clems4ever.

We could still have specific instances where we think forcing a refresh would be useful too. That probably means a scenario with a not so super aggressive TTL and other logic will likely result in sessions being near real-time perfect without too much pressure on LDAP.

On the flipside if someone sets a very aggressive TTL (as in seconds), perhaps we don't need to perform the additional logic, just thinking about how much effect this might have on LDAP and what scenarios we think we should be handling.

**Comment by james-d-elliott** [comment]
Well realistically if there is no rule that matches a request which has a group subject there is no need to update.

So what if we do this @clems4ever @nightah :

1. Implement a new config var under authentication_backend.ldap called ttl as a duration notation type (string in the schema).
2. Have the default set to 5 minutes (blank string)
3. Only check for updated groups if the provider is LDAP and there is a group subject rule that could possibly match
4. If the ttl is set to a string "disable" completely disable the checking.

**Comment by james-d-elliott** [comment]
@abretaud we've agreed on a design. It should be in the next release or two which should be before the weekend.

**Comment by abretaud** [comment]
:tada: thanks a lot!

**Comment by james-d-elliott** [comment]
@abretaud me and @nightah will probably test this over the next day or so:

authelia/authelia:fix-ldap-group-cache

**Comment by james-d-elliott** [comment]
Sorry this has been slightly delayed and will be out with 4.16.1. We had a security flaw we had to fix and wanted a release ASAP.

**Comment by james-d-elliott** [comment]
This has been released in 4.17 @abretaud

**Comment by abretaud** [comment]
Super cool, thanks :)
